src/v/kafka/server/group_metadata_migration.cc

```diff
@@ -0,0 +1,691 @@
+              result.value());
+        }
+
+        co_await ss::sleep_abortable(default_wait_time, as);
+        result = co_await api.get_reconciliation_state(model::kafka_group_nt);
+    }
+}
+
+} // namespace
+
+cluster::feature_table& group_metadata_migration::feature_table() {
+    return _controller.get_feature_table().local();
+}
+
+cluster::feature_manager& group_metadata_migration::feature_manager() {
+    return _controller.get_feature_manager().local();
+}
+ss::abort_source& group_metadata_migration::abort_source() {
+    return _controller.get_abort_source().local();
+}
+
+ss::future<> group_metadata_migration::activate_feature(ss::abort_source& as) {
+    vlog(mlog.info, "activating consumer offsets feature");
+    while (!feature_table().is_active(cluster::feature::consumer_offsets)
+           && !as.abort_requested()) {
+        auto err = co_await feature_manager().write_action(
+          cluster::feature_update_action{
+            .feature_name = ss::sstring(
+              _controller.get_feature_table()
+                .local()
+                .get_state(cluster::feature::consumer_offsets)
+                .spec.name),
+            .action
+            = cluster::feature_update_action::action_t::complete_preparing,
+          });
+        if (
+          err
+          || !feature_table().is_active(cluster::feature::consumer_offsets)) {
+            if (err) {
+                vlog(
+                  mlog.info,
+                  "error activating consumer offsets feature: {}",
+                  err.message());
+            }
+            co_await ss::sleep_abortable(default_timeout, as);
+        }
+    }
+}
+
+ss::future<> group_metadata_migration::do_apply() {
+    vlog(
+      mlog.info,
+      "waiting for consumer offsets feature to be preparing - current "
+      "state: {}",
+      _controller.get_feature_table()
+        .local()
+        .get_state(cluster::feature::consumer_offsets)
+        .get_state());
+    co_await feature_table().await_feature_preparing(
+      cluster::feature::consumer_offsets, abort_source());
+    vlog(mlog.info, "disabling partition movement feature and group router");
+    co_await _group_router.invoke_on_all(&group_router::disable);
+    co_await _controller.get_topics_frontend().invoke_on_all(
+      &cluster::topics_frontend::disable_partition_movement);
+    vlog(mlog.info, "waiting for stable consumer group topic");
+    co_await wait_for_stable_group_topic(
+      _controller.get_api().local(), _controller.get_abort_source().local());
+    vlog(mlog.info, "applying consumer groups migrations");
+    co_await migrate_metadata();
+    vlog(mlog.info, "waiting for all migrations to finish");
+    co_await feature_manager().barrier(active_barrier_tag);
+    vlog(mlog.info, "finishing migration - enabling consumer offsets feature");
+
+    co_await activate_feature(_controller.get_abort_source().local());
+
+    co_await feature_table().await_feature(
```

Comment:
activate_feature also loops until the feature shows as active, so this await_feature may not be needed (or maybe activate_feature should return as soon as it has successfully written, and not wait to see the activation)

Reply:
i found it pretty clear that you first do `activate`, and then `await_active`. I am calling await to wait for the state to be applied to feature table.